execution/executor/src/components/in_memory_state_calculator_v2.rs

```diff
@@ -0,0 +1,253 @@
+    }
+}
+
+impl<'a> AccountView for CoreAccountStateView<'a> {
+    fn get_state_value(&self, state_key: &StateKey) -> Result<Option<Vec<u8>>> {
+        if let Some(v_opt) = self.updates.get(state_key) {
+            return Ok(v_opt.as_ref().map(|x| x.bytes().to_vec()));
+        }
+        if let Some(v_opt) = self.base[state_key.get_shard_id() as usize]
+            .get(state_key)
+            .as_ref()
+        {
+            return Ok(v_opt.as_ref().map(|x| x.bytes().to_vec()));
+        }
+        Ok(None)
+    }
+
+    fn get_account_address(&self) -> Result<Option<AccountAddress>> {
+        Ok(Some(CORE_CODE_ADDRESS))
+    }
+}
+
+/// Helper class for calculating state changes after a block of transactions are executed.
+pub struct InMemoryStateCalculatorV2 {
+    ///// These don't change during the calculation.
+    // This makes sure all in-mem nodes seen while proofs were fetched stays in mem during the
+    // calculation
+    _frozen_base: FrozenSparseMerkleTree<StateValue>,
+    proof_reader: ProofReader,
+
+    // The states at base version.
+    sharded_state_cache: [DashMap<StateKey, Option<StateValue>>; 16],
+    next_version: Version,
+    usage: StateStorageUsage,
+
+    //// These changes whenever make_checkpoint() happens.
+    checkpoint: SparseMerkleTree<StateValue>,
+    checkpoint_version: Option<Version>,
+    // This doesn't need to be frozen since `_frozen_base` holds a ref to the oldest ancestor
+    // already, but frozen SMT is used here anyway to avoid exposing the `batch_update()` interface
+    // on the non-frozen SMT.
+    latest: FrozenSparseMerkleTree<StateValue>,
+}
+
+impl InMemoryStateCalculatorV2 {
+    pub fn new(base: &StateDelta, state_cache: StateCache) -> Self {
+        let StateCache {
+            frozen_base,
+            state_cache,
+            proofs,
+        } = state_cache;
+        // TODO(grao): Only a base tree and version are needed because the current should be the
+        // same as base.
+        let StateDelta {
+            base,
+            base_version,
+            current,
+            current_version,
+            updates_since_base,
+        } = base.clone();
+
+        assert!(base_version == current_version);
+        assert!(updates_since_base.is_empty());
+
+        let sharded_state_cache = arr![DashMap::new(); 16];
+
+        // TODO(grao): Move this logic to an earlier stage.
+        let _: Vec<_> = state_cache
+            .into_iter()
+            .map(|(k, v)| {
+                sharded_state_cache[k.get_shard_id() as usize].insert(k, v);
+            })
+            .collect();
+        Self {
+            _frozen_base: frozen_base,
+            proof_reader: ProofReader::new(proofs),
+
+            sharded_state_cache,
+            next_version: current_version.map_or(0, |v| v + 1),
+            usage: current.usage(),
+
+            checkpoint: base,
+            checkpoint_version: base_version,
+            latest: current.freeze(),
```

Comment:
you are not gonna make multiple checkpoints (or even update `next_version` or anything incrementally), once these are generated, it's already time to return the result, so these fields shouldn't exist at all?

Reply:
True, just too lazy to change the parameter and return value to pass them around among functions, use `self` is easy :)

Reply:
well, that'll make it significantly less confusing, which is the purpose of this, so let's be diligent here :p

Reply:
Done